thrift/binary.go

```diff
@@ -1292,6 +1293,394 @@ func (p *BinaryProtocol) DecodeText(val string, desc *TypeDescriptor, disallowUn
+		}
+		it := vr.MapRange()
+		it.Next()
+		firstKey := it.Key().Interface()
+		firstVal := it.Value().Interface()
+		kt, e := GoType2ThriftType(firstKey)
+		if e != nil {
+			return 0, e
+		}
+		et, e := GoType2ThriftType(firstVal)
+		if e != nil {
+			return 0, e
+		}
+		e = p.WriteMapBegin(kt, et, vr.Len())
+		if e != nil {
+			return 0, e
+		}
+		if _, e := p.WriteAny(firstKey, sliceAsSet); e != nil {
+			return 0, e
+		}
+		if _, e := p.WriteAny(firstVal, sliceAsSet); e != nil {
+			return 0, e
+		}
+		for it.Next() {
+			if _, e := p.WriteAny(it.Key().Interface(), sliceAsSet); e != nil {
+				return 0, e
+			}
+			if _, e := p.WriteAny(it.Value().Interface(), sliceAsSet); e != nil {
+				return 0, e
+			}
+		}
+		return MAP, p.WriteMapEnd()
+	case map[interface{}]interface{}:
+		if len(v) == 0 {
+			return 0, fmt.Errorf("empty map[int]interface{} is not supported")
+		}
+		var firstVal, firstKey interface{}
+		for kk, vv := range v {
+			firstVal = vv
+			firstKey = kk
+			break
+		}
+		kt, e := GoType2ThriftType(firstKey)
+		if e != nil {
+			return 0, e
+		}
+		et, e := GoType2ThriftType(firstVal)
+		if e != nil {
+			return 0, e
+		}
+		e = p.WriteMapBegin(kt, et, len(v))
+		if e != nil {
+			return 0, e
+		}
+		for k, vv := range v {
+			switch kt := k.(type) {
+			case *map[string]interface{}:
+				if _, err := p.WriteAny(*kt, sliceAsSet); err != nil {
+					return 0, err
+				}
+			case *map[int]interface{}:
+				if _, err := p.WriteAny(*kt, sliceAsSet); err != nil {
+					return 0, err
+				}
+			case *map[int8]interface{}:
+				if _, err := p.WriteAny(*kt, sliceAsSet); err != nil {
+					return 0, err
+				}
+			case *map[int16]interface{}:
+				if _, err := p.WriteAny(*kt, sliceAsSet); err != nil {
+					return 0, err
+				}
+			case *map[int32]interface{}:
+				if _, err := p.WriteAny(*kt, sliceAsSet); err != nil {
+					return 0, err
+				}
+			case *map[int64]interface{}:
+				if _, err := p.WriteAny(*kt, sliceAsSet); err != nil {
+					return 0, err
+				}
+			case *map[FieldID]interface{}:
+				if _, err := p.WriteAny(*kt, sliceAsSet); err != nil {
+					return 0, err
+				}
+			case *map[interface{}]interface{}:
+				if _, err := p.WriteAny(*kt, sliceAsSet); err != nil {
+					return 0, err
+				}
+			case *[]interface{}:
+				if _, err := p.WriteAny(*kt, sliceAsSet); err != nil {
+					return 0, err
+				}
+			default:
+				if _, err := p.WriteAny(k, sliceAsSet); err != nil {
+					return 0, err
+				}
+			}
+			if _, e := p.WriteAny(vv, sliceAsSet); e != nil {
+				return 0, e
+			}
+		}
+		return MAP, p.WriteMapEnd()
+	case map[FieldID]interface{}:
+		e := p.WriteStructBegin("")
+		if e != nil {
+			return 0, e
+		}
+		for k, vv := range v {
+			ft, e := GoType2ThriftType(vv)
+			if e != nil {
+				return 0, e
+			}
+			if e := p.WriteFieldBegin("", ft, k); e != nil {
+				return 0, e
+			}
+			if _, e := p.WriteAny(vv, sliceAsSet); e != nil {
+				return 0, e
+			}
+		}
+		return STRUCT, p.WriteFieldStop()
+	default:
```

Comment:
Suggestion: opensource task for the community to encode arbitrary type with reflection, which supports zero-length slice/map.

Reply:
Another approach may be to generate all possible combinations of the types based on `go generate` with a template to achieve best performance.

Reply:
TODO